src/thread.rs

```diff
@@ -225,58 +171,39 @@ pub fn scope<'env, F, R>(f: F) -> thread::Result<R>
 where
     F: FnOnce(&Scope<'env>) -> R,
 {
-    let mut scope = Scope {
+    let scope = Scope {
         joins: RefCell::new(Vec::new()),
-        panics: Vec::new(),
         _marker: PhantomData,
     };
 
-    // Executes the scoped function. Panic will be catched as `Err`.
+    // Execute the scoped function, but catch any panics.
     let result = panic::catch_unwind(panic::AssertUnwindSafe(|| f(&scope)));
-
-    // Joins all the threads.
-    scope.join_all();
-    let panic = scope.panics.pop();
-
-    // If any of the threads panicked, returns the panic's payload.
-    if let Some(payload) = panic {
-        return Err(payload);
-    }
-
-    // Returns the result of the scoped function.
-    result
-}
-
-impl<'env> fmt::Debug for Scope<'env> {
-    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        write!(f, "Scope {{ ... }}")
+    // Join all remaining spawned threads.
+    let mut panics = scope.join_all();
+
+    match result {
+        Ok(res) => if panics.is_empty() {
+            Ok(res)
+        } else {
+            Err(Box::new(panics))
+        },
+        Err(err) => {
+            let mut vec = Vec::with_capacity(1 + panics.len());
+            vec.push(err);
+            vec.append(&mut panics);
+            Err(Box::new(vec))
+        }
```

Comment:
Here's another way of achieving the same thing:

```rust
let mut panics = scope.join_all();

if panics.is_empty() {
    result.map_err(|err| Box::new(vec![err]))
} else {
    if let Err(err) = result {
        panics.reserve(1);
        panics.insert(0, err);
    }
    Err(Box::new(panics))
}
```

But I'm okay with both, pick whatever you like best :)

Reply:
> ```rust
> let mut panics = scope.join_all();
> 
> if panics.is_empty() {
>     result.map_err(|err| Box::new(vec![err]))
> } else {
>     if let Err(err) = result {
>         panics.reserve(1);
>         panics.insert(0, err);
>     }
>     Err(Box::new(panics))
> }
> ```

I had tried something similar as well, but for some reason the compiler (latest stable) does not like the `map_err()` part.

```
error[E0308]: mismatched types
   --> src/thread.rs:185:9
    |
170 | pub fn scope<'env, F, R>(f: F) -> thread::Result<R>
    |                                   ----------------- expected `std::result::Result<R, std::boxed::Box<(dyn std::any::Any + std::marker::Send + 'static)>>` because of return type
...
185 |         result.map_err(|err| Box::new(vec![err]))
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected trait std::any::Any, found struct `std::vec::Vec`
    |
    = note: expected type `std::result::Result<_, std::boxed::Box<(dyn std::any::Any + std::marker::Send + 'static)>>`
               found type `std::result::Result<_, std::boxed::Box<std::vec::Vec<std::boxed::Box<dyn std::any::Any + std::marker::Send>>>>`
``` 
It doesn't realize that `Vec<Box<Any + Send + 'static>` fulfils the required trait bounds.
Since it compiles fine with the manual match expressions, this might even be a bug.

Reply:
So the compiler can coerce `Box<Vec<T>>` into `Box<Any + Send>`, but cannot coerce `Result<R, Box<Vec<T>>` into `Result<R, Box<Any + Send>>`.

We need to coerce `Box<Vec<T>>` into `Box<Any + Send>` before we get into a `Result<R, Box<Vec<T>>`:

```rust
result.map_err(|err| Box::new(vec![err]) as _)
```

This will compile.

Reply:
Huh, when in doubt just `as _`.

Thank you for the tip. When this PR is through, I have some things in the pipeline for stats counting as well, but I wanted to base that on the reformatted code for ease of use. It might also make sense to promote `mod thread` to its own directory for this.